plugins/jobtype/src/azkaban/jobtype/HadoopSecureSparkWrapper.java

```diff
@@ -62,47 +86,142 @@ public static void main(final String[] args) throws Exception {
+    logger.info("Args before adjusting driver java opts: " + Arrays.toString(newArgs));
+
+    // Adjust driver java opts param
+    handleDriverJavaOpts(newArgs);
+
+    // If dynamic allocation policy for this jobtype is turned on, adjust related param
+    handleDynamicResourceAllocation(newArgs);
+
+    // If yarn cluster enables node labeling, adjust related param
+    handleNodeLabeling(newArgs);
 
-    final String[] newArgs = concat.toString().split(SparkJobArg.delimiter);
-    logger.info("newArgs: " + Arrays.toString(newArgs));
+    // Realign params after adjustment
+    newArgs = removeNullsFromArgArray(newArgs);
+    logger.info("Args after adjusting driver java opts: " + Arrays.toString(newArgs));
+
+    org.apache.spark.deploy.SparkSubmit$.MODULE$.main(newArgs);
+  }
+
+  private static void handleDriverJavaOpts(String[] argArray) {
+    Configuration conf = new Configuration();
+    // Driver java opts is always the first elem(param name) and second elem(value) in the argArray
+    // Get current driver java opts here
+    StringBuilder driverJavaOptions = new StringBuilder(argArray[1]);
+    // In spark-submit, when both --driver-java-options and conf spark.driver.extraJavaOptions is used,
+    // spark-submit will only pick --driver-java-options, an arg we always have
+    // So if user gives --conf spark.driver.extraJavaOptions=XX, we append the value in --driver-java-options
+    for (int i = 0; i < argArray.length; i++) {
+      if (argArray[i].equals(SparkJobArg.SPARK_CONF_PREFIX.sparkParamName) 
+        && argArray[i+1].startsWith(SPARK_CONF_EXTRA_DRIVER_OPTIONS)) {
+        driverJavaOptions.append(argArray[++i].substring(SPARK_CONF_EXTRA_DRIVER_OPTIONS.length() + 1));
+      }
+    }
     
-    StringBuilder driverJavaOptions = new StringBuilder(newArgs[1]);    
+    // Append addtional driver java opts about azkaban context
     String[] requiredJavaOpts = { WORKFLOW_LINK, JOB_LINK, EXECUTION_LINK, ATTEMPT_LINK };
     for (int i = 0; i < requiredJavaOpts.length; i++) {
-    	  driverJavaOptions.append(" ").append(HadoopJobUtils.javaOptStringFromHadoopConfiguration(conf,
-    	            requiredJavaOpts[i]));
+        driverJavaOptions.append(" ").append(HadoopJobUtils.javaOptStringFromHadoopConfiguration(conf,
+                  requiredJavaOpts[i]));
     }
-    newArgs[1] = driverJavaOptions.toString();
-    logger.info("newArgs2: " + Arrays.toString(newArgs));
+    // Update driver java opts
+    argArray[1] = driverJavaOptions.toString();
+  }
 
-    org.apache.spark.deploy.SparkSubmit$.MODULE$.main(newArgs);
+  private static void handleDynamicResourceAllocation(String[] argArray) {
+    // HadoopSparkJob will set env var on this process if we enable dynamic allocation policy for spark jobtype.
+    // This policy can be enabled in spark jobtype plugin's conf property.
+    // This policy is designed to enforce dynamic allocation.
```

Comment:
Better to further document that enabling dynamic allocation policy for spark job type is different from enabling dynamic allocation feature for Spark. This config inside Spark job type is to enforce dynamic allocation feature for all Spark applications submitted via Azkaban Spark job type.